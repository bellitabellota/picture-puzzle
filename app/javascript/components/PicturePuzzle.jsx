import React, {useRef, useState} from "react";
import { Link } from "react-router-dom";
import IncorrectMessage from "./IncorrectMessage";
import SelectBoxContainer from "./SelectBoxContainer";

//The puzzle will be retrieved from the database or passed in in the parent as prop?

const puzzle = {
  id: 1,
  title: "The Smurfs",
  imageSrc: "/picture-puzzle-images/the-smurfs.jpg",
  taskDescription: "Find Papa Smurf, Smurfette and Brainy Smurf.",
  resolution: [1000, 674],
  targets: [
    { name: "Smurfette", boundingBox: { xMin: 90, xMax: 140, yMin: 360, yMax: 440 }},
    { name: "Papa Smurf", boundingBox: { xMin: 410, xMax: 475, yMin: 420, yMax: 507 }}, 
    { name: "Brainy Smurf", boundingBox: { xMin: 277, xMax: 343, yMin: 525, yMax: 655}}
  ]
}

function PicturePuzzle() {
  const [clickedCoordinates, setClickedCoordinates] = useState({ });
  const [correctlyIdentifiedTargets, setCorrectlyIdentifiedTargets] = useState([]);
  const [incorrectMessage, setIncorrectMessage] = useState(null);

  // Finish and call the below placeCorrectSymbols function
  function placeCorrectSymbols(correctlyIdentifiedTargets) {

  }

  //Call function that checks if correctlyIdentifiedTargets matches targets to see if puzzle completed.

  function getCoordinates(event) {
    const img = event.target;
    const rect = img.getBoundingClientRect();

    const displayedWidth = rect.width;
    const displayedHeight = rect.height;

    const scalingFactorX = displayedWidth / puzzle.resolution[0];
    const scalingFactorY = displayedHeight / puzzle.resolution[1];

    // Calculate relative click position
    const scaledX = event.clientX - rect.left;
    const scaledY = event.clientY - rect.top;

    // Scale click coordinates back to original resolution
    const originalX = Math.round(scaledX / scalingFactorX);
    const originalY = Math.round(scaledY / scalingFactorY);

    setClickedCoordinates({originalX:originalX, originalY: originalY, scaledX: scaledX, scaledY:scaledY, isSelecting: true})
  }

  function getTargetOfClickedCoordinates(originalX, originalY) {
    let clickedTarget = null;

    puzzle.targets.forEach((target) => {
      if (isInsideBoundingBox(originalX, originalY, target.boundingBox)) {
        clickedTarget = target;
      }
    })
    return clickedTarget;
  }

  function isInsideBoundingBox(x, y, box) {
    return x >= box.xMin && x <= box.xMax && y >= box.yMin && y <= box.yMax;
  }

  const selectBox = useRef(null);

  function confirmTargetSelection() {
    const selectedTargetName = selectBox.current.value;
    const clickedTarget = getTargetOfClickedCoordinates(clickedCoordinates.originalX, clickedCoordinates.originalY);

    if (clickedTarget && (selectedTargetName === clickedTarget.name)) {
      if (!correctlyIdentifiedTargets.includes(clickedTarget)) {
        // Add to correctlyIdentifiedTargets state variable
        setCorrectlyIdentifiedTargets([...correctlyIdentifiedTargets, clickedTarget]);
      } else {
        setIncorrectMessage("Already identified. Find the remaining targets.")
      }

    } else {
      setIncorrectMessage("Not quite. Try it again.")
    }

    // Ensure that the select box is removed from screen (until next click on image) 
    setClickedCoordinates({...clickedCoordinates, isSelecting: false})
  }

  return (
    <main className="main-picture-puzzle">
      {incorrectMessage && <IncorrectMessage message={incorrectMessage} setIncorrectMessage={setIncorrectMessage} />}
      <Link to="/" >&lt; Back to Home</Link>
      <h1>{puzzle.title}</h1>
      <div className="task-info">
        <p>{puzzle.taskDescription}</p>
        <Timer />
      </div>

      <div className="img-container">
        {clickedCoordinates.isSelecting && 
          <SelectBoxContainer clickedCoordinates={clickedCoordinates} selectBox={selectBox} targets={puzzle.targets} confirmTargetSelection={confirmTargetSelection} />
        }
        
        <img src={puzzle.imageSrc} onClick={getCoordinates} />
      </div>
    </main>
  )
}

export default PicturePuzzle;

function Timer() {
  return(
    <p className="timer">00:00</p>
  )
}
