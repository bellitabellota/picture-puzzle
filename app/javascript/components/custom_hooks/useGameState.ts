import {useEffect, useState} from "react";

import type { PuzzleType } from "../../types/PuzzleType";
import type { IdentifiedTargetType } from "../../types/IdentifiedTargetType";

const useGameState = (puzzle: PuzzleType, correctlyIdentifiedTargets: IdentifiedTargetType[], paramsId: string) => {
  console.log(puzzle)
  
  const [secondsToCompletion, setSecondsToCompletion] = useState(null);
  const [gameStateError, setGameStateError] = useState(null);

  useEffect(() => {
    if (puzzle && (correctlyIdentifiedTargets.length === puzzle.targets.length)) {
      const url = `/api/v1/puzzle_validations/${paramsId}/game_state`

      fetch(url)
      .then((response) => {
        if(!response.ok) {
          throw new Error(`HTTP Error ${response.status}: ${response.statusText}`);
        }
        return response.json()
      }).then((data) => {
        if (data.gameFinished === true) {
          setSecondsToCompletion(data.secondsToCompletion);
        }
      }).catch(error => setGameStateError(error));
    }
  }, [correctlyIdentifiedTargets])

  return {secondsToCompletion, gameStateError}
}

export default useGameState;