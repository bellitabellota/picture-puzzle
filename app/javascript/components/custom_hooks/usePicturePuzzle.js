import {useState, useEffect} from "react";

const usePicturePuzzle = (paramsId) => {
  const [puzzle , setPuzzle ] = useState(null);
  const [error, setError] = useState(null);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    const controller = new AbortController();
    const signal = controller.signal;
    
    const url = `/api/v1/picture_puzzles/${paramsId}`
    
    fetch(url, {signal})
    .then((response) => {
      if(!response.ok) {
        throw new Error("Network response was not ok.");
      }
      return response.json();
    })
    .then((response) => {
      setPuzzle({...response,
        imageSrc: response.image_src,
        taskDescription: response.task_description
      });
    }).catch((error) => {
      if (error.name !== "AbortError") {
        setError(error);
      }
    }).finally(() => {
      if (!signal.aborted) {
        setIsLoading(false);
      }
    })

    return () => controller.abort(); // the cleanup function was added because the effect runs when the component mounts. So in case the component unmount before the request completes, the request would have been still active but the component would not be there anymore to handle the response.

  }, [paramsId])

  return { puzzle, error, isLoading }
}

export default usePicturePuzzle;