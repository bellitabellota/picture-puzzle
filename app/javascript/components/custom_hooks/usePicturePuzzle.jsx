import React, {useState, useEffect, use} from "react";

const usePicturePuzzle = (paramsId) => {
  const [puzzle , setPuzzle ] = useState(null);
  const [error, setError] = useState(null);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    const controller = new AbortController();
    const signal = controller.signal;
    
    const url = `/api/v1/picture_puzzles/${paramsId}`
    
    fetch(url, {signal})
    .then((response) => {
      if(!response.ok) {
        throw new Error("Network response was not ok.");
      }
      return response.json();
    })
    .then((response) => {
      setPuzzle({...response,
        imageSrc: response.image_src,
        taskDescription: response.task_description
      });
    }).catch((error) => {
      if (error.name !== "AbortError") {
        setError(error);
      }
    }).finally(() => {
      if (!signal.aborted) {
        setIsLoading(false);
      }
    })

    return () => controller.abort();

  }, [paramsId])

  return { puzzle, error, isLoading }
}

export default usePicturePuzzle;