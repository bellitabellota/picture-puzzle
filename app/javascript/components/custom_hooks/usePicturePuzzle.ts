import {useState, useEffect} from "react";

import type { PuzzleType } from "../../types/PuzzleType";
import type { TargetNameObjectType } from "../../types/TargetNameObjectType";

function isTargetNameObject(data: any): data is TargetNameObjectType {
  return (
    data &&
    typeof data === "object" &&
    typeof data.name === "string"
  )
}

function isPuzzle(data:any): data is PuzzleType {
  return(
    data &&
    typeof data.id === "number" &&
    typeof data.title === "string" &&
    typeof data.imageSrc === "string" &&
    typeof data.taskDescription=== "string" &&
    typeof data.resolutionWidth === "number" &&
    typeof data.resolutionHeight === "number" &&
    Array.isArray(data.targets) && data.targets.every(isTargetNameObject) &&
    typeof data.createdAt === "string" &&
    typeof data.updatedAt === "string"
  )
}

const usePicturePuzzle = (paramsId: string) => {
  const [puzzle , setPuzzle ] = useState<PuzzleType|null>(null);
  const [error, setError] = useState<Error|null>(null);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    const controller = new AbortController();
    const signal = controller.signal;
    
    const url = `/api/v1/picture_puzzles/${paramsId}`
    
    fetch(url, {signal})
    .then((response) => {
      if(!response.ok) {
        throw new Error(`HTTP Error ${response.status}: ${response.statusText}`);
      }
      return response.json();
    })
    .then((response) => {
      if (!isPuzzle(response)) {
        throw Error("Invalid puzzle data from API.")
      }
      setPuzzle(response);
    }).catch((error) => {
      if (error.name !== "AbortError") {
        setError(error);
      }
    }).finally(() => {
      if (!signal.aborted) {
        setIsLoading(false);
      }
    })

    return () => controller.abort(); // the cleanup function was added because the effect runs when the component mounts. So in case the component unmount before the request completes, the request would have been still active but the component would not be there anymore to handle the response.

  }, [paramsId])

  return { puzzle, error, isLoading }
}

export default usePicturePuzzle;