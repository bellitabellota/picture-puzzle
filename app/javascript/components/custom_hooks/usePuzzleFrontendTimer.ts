
import {useState, useEffect, useRef} from "react";

const usePuzzleFrontendTimer = (puzzle, secondsToCompletion) => {
  const [secondsPassed, setSecondsPassed] = useState(0);
  const savedIncreaseSecondsPassed = useRef(null);

  function increaseSecondsPassed() {
    setSecondsPassed(secondsPassed + 1);
  }

  useEffect(()=> {
    // saving the increaseSecondsPassed function on every render via useRef 
    // makes sure that when the function is passed into setInterval in the Effect below 
    // always the most recent state of secondsPassed is referenced
    savedIncreaseSecondsPassed.current = increaseSecondsPassed;
  })

  useEffect(()=>{
    let id;
    if (secondsToCompletion) return;

    if (puzzle) {
      id = setInterval(() => {savedIncreaseSecondsPassed.current()},  1000)
    }

    return () => {
      if (id) {
        clearInterval(id);
      }
    }
  }, [puzzle, secondsToCompletion])

  return { secondsPassed }
}

export default usePuzzleFrontendTimer ;