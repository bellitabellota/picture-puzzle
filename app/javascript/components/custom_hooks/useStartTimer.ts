import {useEffect, useState} from "react";

const useStartTimer = (puzzle, paramsId) => {
  const [startTimerError, setStartTimerError] = useState(null); 

  useEffect(()=> {
    if(puzzle ) {
      const url = `/api/v1/puzzle_timers/${paramsId}/start_timer`;
      const token = document.querySelector<HTMLMetaElement>('meta[name="csrf-token"]').content;

      fetch(url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "X-CSRF-Token": token
        }
      })
      .then((response) => {
        if (!response.ok) {
          return response.json().then((errorData) => {
            throw new Error(errorData.error || `HTTP Error ${response.status}: ${response.statusText}`);
          });
        }
      })
      .catch((error) => setStartTimerError(error));
    }
  }, [puzzle])

  return {startTimerError}
}

export default useStartTimer;