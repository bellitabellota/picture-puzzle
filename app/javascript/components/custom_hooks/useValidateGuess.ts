import {useEffect, useState} from "react";

const useValidateGuess = (selectedName, setSelectedName, paramsId, clickedCoordinates, incorrectMessage, setIncorrectMessage) => {
  const [correctlyIdentifiedTargets, setCorrectlyIdentifiedTargets] = useState([]);
  const [validationError, setValidationError] = useState(null);

  useEffect(()=>{
    if (!selectedName) return;

    const url = `/api/v1/puzzle_validations/${paramsId}/validate_guess`
    const token = document.querySelector('meta[name="csrf-token"]').content;
    const body = {originalX: clickedCoordinates.originalX, originalY: clickedCoordinates.originalY, selectedName}

    fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-CSRF-Token": token
      },
      body: JSON.stringify(body),
    }).then((response) => {

      if (!response.ok) {
        return response.json().then((errorData) => {
          throw new Error(errorData.error || `HTTP Error ${response.status}: ${response.statusText}`);
        });
      }

      return response.json()

    }).then((data) => {
      if (data.success === true && !correctlyIdentifiedTargets.includes(data.target)) {
        setCorrectlyIdentifiedTargets([...correctlyIdentifiedTargets, data.target]);
      } else {
        setIncorrectMessage(data.message);
      }
      setSelectedName(null);
    }).catch(error => { setValidationError(error)})
  }, [selectedName])

  return {incorrectMessage, correctlyIdentifiedTargets, validationError}
}

export default useValidateGuess;