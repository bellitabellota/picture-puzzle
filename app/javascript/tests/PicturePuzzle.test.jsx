import PicturePuzzle from "../components/PicturePuzzle";
import usePicturePuzzle from "../components/custom_hooks/usePicturePuzzle";
import {render, screen, fireEvent } from "@testing-library/react";
import React from "react";
import { createMemoryRouter, RouterProvider } from "react-router-dom";
import routes from "../routes/index"
import useValidateGuess from "../components/custom_hooks/useValidateGuess";
import useGameState from "../components/custom_hooks/useGameState";
import useStartTimer from "../components/custom_hooks/useStartTimer";
import usePuzzleFrontendTimer from "../components/custom_hooks/usePuzzleFrontendTimer";

vi.mock("../components/custom_hooks/usePicturePuzzle", () => ({ default: vi.fn() }))
vi.mock("../components/custom_hooks/useValidateGuess", () => ({ default: vi.fn() }))
vi.mock("../components/custom_hooks/useGameState", () => ({ default: vi.fn() }))
vi.mock("../components/custom_hooks/useStartTimer", () => ({ default: vi.fn() }))
vi.mock("../components/custom_hooks/usePuzzleFrontendTimer", () => ({ default: vi.fn() }))

const testPuzzle = {
  id: 123,
  title: "Fun Test Puzzle",
  imageSrc: "/picture-puzzle-images/test-puzzle.jpg",
  taskDescription: "Find all Test Targets.",
  resolution_width: 1280,
  resolution_height: 720,
  targets: [
      { name: "Test Target 1" },
      { name: "Test Target 2" },
      { name: "Test Target 3" }
  ]
}

const memoryRouter = createMemoryRouter(routes, { initialEntries: ["/123"] })

describe("PicturePuzzle", ()=> {
  it("renders 'Puzzle is loading ...' when isLoading is true", () => {
    usePicturePuzzle.mockReturnValue({puzzle: null, error: null, isLoading: true})
    useValidateGuess.mockReturnValue({ correctlyIdentifiedTargets: [], validationError: null });
    useGameState.mockReturnValue({ secondsToCompletion: null, gameStateError: null })
    useStartTimer.mockReturnValue({ startTimerError: null });

    usePuzzleFrontendTimer.mockReturnValue({ secondsPassed: 0 });

    render(<PicturePuzzle />)

    expect(screen.getByText("Puzzle is loading ...")).toBeInTheDocument();
  })

  it("renders error message if usePicturePuzzle returned an Error", () => {
    usePicturePuzzle.mockReturnValue({puzzle: null, error: new Error("server error"), isLoading: false})
    useValidateGuess.mockReturnValue({ correctlyIdentifiedTargets: [], validationError: null });
    useGameState.mockReturnValue({ secondsToCompletion: null, gameStateError: null })
    useStartTimer.mockReturnValue({ startTimerError: null });

    usePuzzleFrontendTimer.mockReturnValue({ secondsPassed: 0 });

    render(<PicturePuzzle />)
    expect(screen.getByText("server error")).toBeInTheDocument();
  })

  it("renders error message if useStartTimer returned an Error", () => {
    usePicturePuzzle.mockReturnValue({puzzle: null, error: null, isLoading: false})
    useValidateGuess.mockReturnValue({ correctlyIdentifiedTargets: [], validationError: null });
    useGameState.mockReturnValue({ secondsToCompletion: null, gameStateError: null })
    useStartTimer.mockReturnValue({ startTimerError: new Error("useStartTimer error") });

    usePuzzleFrontendTimer.mockReturnValue({ secondsPassed: 0 });

    render(<PicturePuzzle />)
    expect(screen.getByText("useStartTimer error - The server could not correctly set the start time.")).toBeInTheDocument();
  })

  it("renders puzzle if usePicturePuzzle returned a puzzle", ()=> {
    usePicturePuzzle.mockReturnValue({ puzzle: testPuzzle, error: false, isLoading: false})
    useValidateGuess.mockReturnValue({ correctlyIdentifiedTargets: [], validationError: null });
    useGameState.mockReturnValue({ secondsToCompletion: null, gameStateError: null })
    useStartTimer.mockReturnValue({ startTimerError: null });

    usePuzzleFrontendTimer.mockReturnValue({ secondsPassed: 0 });

    const {container} = render(
      <RouterProvider router={memoryRouter}></RouterProvider>
    )

    expect(container).toMatchSnapshot();
  })

  it('shows the target selection box when image is clicked', () => {
    usePicturePuzzle.mockReturnValue({ puzzle: testPuzzle, error: false, isLoading: false})
    useValidateGuess.mockReturnValue({ correctlyIdentifiedTargets: [], validationError: null });
    useGameState.mockReturnValue({ secondsToCompletion: null, gameStateError: null })
    useStartTimer.mockReturnValue({ startTimerError: null });
    usePuzzleFrontendTimer.mockReturnValue({ secondsPassed: 0 });
  
    render(
      <RouterProvider router={memoryRouter}></RouterProvider>
    )
    
    const img = screen.getByRole('img');
    fireEvent.click(img, { clientX: 100, clientY: 100 });
    
    expect(screen.getByRole('combobox')).toBeInTheDocument();
  });

  it('displays check mark when target is correctly identified', () => {
    usePicturePuzzle.mockReturnValue({ puzzle: testPuzzle, error: false, isLoading: false})
    useValidateGuess.mockReturnValue({ correctlyIdentifiedTargets: [{ name: "Identified Test Target", xCenter: 100, yCenter: 61 }], validationError: null });
    useGameState.mockReturnValue({ secondsToCompletion: null, gameStateError: null })
    useStartTimer.mockReturnValue({ startTimerError: null });
    usePuzzleFrontendTimer.mockReturnValue({ secondsPassed: 0 });
    
    render(
      <RouterProvider router={memoryRouter}></RouterProvider>
    );
    
    const checkMarks = screen.getAllByRole('img', { name: '' });
    const checkMark = checkMarks.find(img => img.className.includes('check-mark'));
    
    expect(checkMark).toBeInTheDocument();
  })

  it('shows completion modal when puzzle is completed', () => {
    usePicturePuzzle.mockReturnValue({ puzzle: testPuzzle, error: false, isLoading: false})
    useValidateGuess.mockReturnValue({ correctlyIdentifiedTargets: [], validationError: null });
    useGameState.mockReturnValue({ secondsToCompletion: 91, gameStateError: null })
    useStartTimer.mockReturnValue({ startTimerError: null });
    usePuzzleFrontendTimer.mockReturnValue({ secondsPassed: 0 });
    
    render(
      <RouterProvider router={memoryRouter}></RouterProvider>
    );
    
    expect(screen.getByText("You solved the puzzle in 1 minute(s) and 31 seconds.")).toBeInTheDocument();
  })
});