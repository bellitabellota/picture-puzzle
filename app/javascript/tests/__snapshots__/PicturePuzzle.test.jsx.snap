// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PicturePuzzle > renders puzzle if usePicturePuzzle returned a puzzle 1`] = `
<div>
  <main
    class="main-picture-puzzle"
  >
    <a
      data-discover="true"
      href="/"
    >
      &lt; Back to Home
    </a>
    <h1>
      Fun Test Puzzle
    </h1>
    <div
      class="task-info"
    >
      <p>
        Find all Test Targets.
      </p>
      <p
        class="timer"
      >
        00:00
      </p>
    </div>
    <div
      class="img-container"
    >
      <img
        src="/picture-puzzle-images/test-puzzle.jpg"
      />
    </div>
  </main>
</div>
`;
